scs2-definition/src/main/java/us/ihmc/scs2/definition/robot/urdf/URDFTools.java

```diff
@@ -209,66 +214,188 @@ public static URDFModel loadURDFModel(InputStream inputStream, Collection<String
+
+            if (!parserProperties.parseSensors && urdfModel.getGazebos() != null)
+            {
+               urdfModel.getGazebos().removeIf(gazebo -> gazebo.getSensor() != null);
+            }
+
+            if (parserProperties.handleImplicitJointDefinitions)
+               handleImplicitJointDefinitions(urdfModel);
+
+            // Merge the current URDFModel with the combined URDFModel
+            combinedURDFModel = mergeURDFModels(combinedURDFModel, urdfModel);
          }
-         else
+      }
+      finally
+      {
+         for (InputStream inputStream : inputStreams)
          {
-            InputSource is = new InputSource(inputStream);
-            SAXParserFactory sax = SAXParserFactory.newInstance();
-            sax.setNamespaceAware(false);
-            XMLReader reader;
-
             try
             {
-               reader = sax.newSAXParser().getXMLReader();
+               inputStream.close();
             }
-            catch (SAXException | ParserConfigurationException e)
+            catch (IOException e)
             {
-               throw new JAXBException(e);
+               LogTools.error(e.getMessage());
             }
-
-            SAXSource source = new SAXSource(reader, is);
-            urdfModel = (URDFModel) um.unmarshal(source);
          }
-         resolvePaths(urdfModel, allResourceDirectories, resourceClassLoader);
+      }
 
-         if (!parserProperties.linksToIgnore.isEmpty() && urdfModel.getLinks() != null)
-            urdfModel.getLinks().removeIf(urdfLink -> parserProperties.linksToIgnore.contains(urdfLink.getName()));
-         if (!parserProperties.jointsToIgnore.isEmpty() && urdfModel.getJoints() != null)
-            urdfModel.getJoints().removeIf(urdfJoint -> parserProperties.jointsToIgnore.contains(urdfJoint.getName()));
-         if (!parserProperties.parseSensors)
-            urdfModel.getGazebos().removeIf(gazebo -> gazebo.getSensor() != null);
+      return combinedURDFModel;
+   }
 
-         if (parserProperties.handleImplicitJointDefinitions)
-            handleImplicitJointDefinitions(urdfModel);
+   private static URDFModel mergeURDFModels(URDFModel baseModel, URDFModel additionalModel)
+   {
+      // Create a new URDFModel to hold the merged components
+      URDFModel mergedModel = new URDFModel();
+      List<String> mergedLinkNames = new ArrayList<>();
+      List<String> mergedJointNames = new ArrayList<>();
+      List<String> mergedGazeboNames = new ArrayList<>();
 
-         return urdfModel;
+      // Set name of merged urdf
+      if (baseModel.getName() == null)
+      {
+         mergedModel.setName(additionalModel.getName());
       }
-      finally
+      else
       {
-         try
+         mergedModel.setName(baseModel.getName());
+      }
+
+      // Merge links that haven't already been added
+      List<URDFLink> mergedLinks = new ArrayList<>();
+      if (baseModel.getLinks() != null)
+      {
+         mergedLinks.addAll(baseModel.getLinks());
```

Comment:
Add all the links in the baseModel, record the names of them as well to avoid duplicates later.